[BE][Attention] Factor out common code

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139791
* #139763
* #139784
* __->__ #139788

- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

malfet committed:
[BE][Attention] Factor out common code

- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

[ghstack-poisoned]

malfet committed:
Update on "[BE][Attention] Factor out common code"


- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

[ghstack-poisoned]

malfet committed:
Update on "[BE][Attention] Factor out common code"


- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -703,30 +703,32 @@ Tensor scaled_dot_product_attention(
     bool is_causal,
     std::optional<double> scale,
     bool enable_gqa) {
+  using namespace sdp;
```
personally dont love maybe just bring in the two objects in the namespace

drisspg commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -745,18 +746,19 @@ Tensor scaled_dot_product_attention(
-          && query_.is_contiguous() && key.is_contiguous() && value.is_contiguous()
-          && !query_.is_nested() && !key.is_nested() && !value.is_nested()) {
+    case SDPBackend::math: {
+      const auto any_nested = query_.is_nested() || key.is_nested() || value.is_nested();
```
So these checks should ideally be in the dispatching logic, I am not sure if MPS currently has a stub for doing the input validation. I think that it this length of check its okay but if it gets any larger I think we should do it properly

malfet committed:
Update on "[BE][Attention] Factor out common code"


- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

[ghstack-poisoned]

malfet committed:
Update on "[BE][Attention] Factor out common code"


- Compute attention mask before the switch
- Introduce `query_device_type` variable
- Refactor some of MPS-math checks into easily readable boolean names

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "This supposed to be a no-op"
